File: app/src/main/java/de/herrmann_engel/rbv/utils/SearchCards.java
package de.herrmann_engel.rbv.utils;

import static de.herrmann_engel.rbv.Globals.LIST_ACCURATE_SIZE;

import java.util.List;
import java.util.regex.Pattern;

import de.herrmann_engel.rbv.db.DB_Card;

public class SearchCards {

    private boolean compare(String source, String query) {
        if (source == null) {
            return false;
        }
        return Pattern.compile(Pattern.quote(query), Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE).matcher(source)
                .find();
    }

    private boolean compareInaccurate(String source, String query) {
        if (source == null) {
            return false;
        }
        return source.toLowerCase().contains(query);
    }

    public List<DB_Card> searchCards(List<DB_Card> input, String query) {
        if (input.size() > LIST_ACCURATE_SIZE) {
            String queryLower = query.toLowerCase();
            input.removeIf(l -> !compareInaccurate(l.front, queryLower) && !compareInaccurate(l.back, queryLower) && !compareInaccurate(l.notes, queryLower));
        } else {
            input.removeIf(l -> !compare(l.front, query) && !compare(l.back, query) && !compare(l.notes, query));
        }
        return input;
    }
}

File: app/src/main/java/de/herrmann_engel/rbv/utils/SearchCards.kt
package de.herrmann_engel.rbv.utils

import de.herrmann_engel.rbv.Globals
import de.herrmann_engel.rbv.db.DB_Card_With_Meta
import java.util.Locale
import java.util.regex.Pattern

class SearchCards {
    private fun hasNoMatch(source: String?, query: String): Boolean {
        return if (source == null) {
            true
        } else !Pattern.compile(
            Pattern.quote(query),
            Pattern.CASE_INSENSITIVE or Pattern.UNICODE_CASE
        ).matcher(source)
            .find()
    }

    private fun hasNoMatchInaccurate(source: String?, query: String): Boolean {
        return if (source == null) {
            true
        } else !source.lowercase(Locale.ROOT).contains(query)
    }

    fun searchCards(input: MutableList<DB_Card_With_Meta>, query: String) {
        if (input.size > Globals.LIST_ACCURATE_SIZE) {
            val queryLower = query.lowercase(Locale.ROOT)
            input.removeIf { l: DB_Card_With_Meta ->
                hasNoMatchInaccurate(
                    l.formattedFront ?: l.card.front,
                    queryLower
                ) && hasNoMatchInaccurate(
                    l.formattedBack ?: l.card.back,
                    queryLower
                ) && hasNoMatchInaccurate(
                    l.card.notes,
                    queryLower
                ) && (l.tagNames.isNullOrBlank() || hasNoMatchInaccurate(
                    l.tagNames,
                    queryLower
                ))
            }
        } else {
            input.removeIf { l: DB_Card_With_Meta ->
                hasNoMatch(
                    l.formattedFront ?: l.card.front,
                    query
                ) && hasNoMatch(
                    l.formattedBack ?: l.card.back,
                    query
                ) && hasNoMatch(
                    l.card.notes,
                    query
                ) && (l.tagNames.isNullOrBlank() || hasNoMatch(
                    l.tagNames,
                    query
                ))
            }
        }
    }
}
